core/src/main/java/com/backblaze/b2/client/B2LargeFileStorer.java

```diff
@@ -158,39 +227,156 @@ private B2FileVersion finishLargeFileFromB2PartFutures(B2FileVersion largeFileVe
+                    uploadListener,
+                    partNumber,
+                    contentSource.getContentLength(),
+                    0,
+                    B2UploadState.FAILED);
+
+            throw e;
+        }
     }
 
-    B2Part copyPart(int partNumber, String sourceFileId, B2ByteRange byteRangeOrNull) throws B2Exception {
+    /**
+     * Stores a part by copying from a file that is already stored in a bucket.
+     *
+     * We do not know the true size of the part until it is finally stored. Some
+     * copy operations do not provide a byte range being copied, and byte ranges
+     * can be clamped down if they exceed the bounds of the file. Therefore, we
+     * use a placeholder value until the operation succeeds. Once the API returns
+     * a B2Part object, we supply the true size in the SUCCEEDED event.
+     */
+    B2Part copyPart(
+            int partNumber,
+            String sourceFileId,
+            B2ByteRange byteRangeOrNull,
+            B2UploadListener uploadListener) throws B2Exception {
+
+        updateProgress(
+                uploadListener,
+                partNumber,
+                B2UploadProgress.UNKNOWN_PART_SIZE_PLACEHOLDER,
+                0,
+                B2UploadState.WAITING_TO_START);
+
         final B2CopyPartRequest copyPartRequest = B2CopyPartRequest
                 .builder(partNumber, sourceFileId, fileVersion.getFileId())
                 .setRange(byteRangeOrNull)
                 .build();
 
-        return retryer.doRetry(
-                "b2_copy_part",
-                accountAuthCache, () -> webifier.copyPart(accountAuthCache.get(), copyPartRequest),
-                retryPolicySupplier.get());
+        try {
+            return retryer.doRetry(
+                    "b2_copy_part",
+                    accountAuthCache,
+                    () -> {
+                        updateProgress(
+                                uploadListener,
+                                partNumber,
+                                B2UploadProgress.UNKNOWN_PART_SIZE_PLACEHOLDER,
+                                0,
+                                B2UploadState.STARTING);
+
+                        final B2Part part = webifier.copyPart(accountAuthCache.get(), copyPartRequest);
+
+                        updateProgress(
+                                uploadListener,
+                                partNumber,
+                                part.getContentLength(),
+                                part.getContentLength(),
+                                B2UploadState.SUCCEEDED);
+
+                        return part;
+                    },
+                    retryPolicySupplier.get());
+        } catch (Exception e) {
+            updateProgress(
+                    uploadListener,
+                    partNumber,
+                    B2UploadProgress.UNKNOWN_PART_SIZE_PLACEHOLDER,
+                    0,
+                    B2UploadState.FAILED);
+
+            throw e;
```

Comment:
How can you throw `Exception` here when the method throws the more specific `B2Exception`?

Reply:
Not sure why I was catching Exception - switched to B2Exception.